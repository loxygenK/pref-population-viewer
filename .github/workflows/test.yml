name: "Test"

on:
  pull_request:
    types:
      - opened
      - synchronize

jobs:
  check:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - uses: actions/cache@v2
        with:
          path: ~/.pnpm-store
          key: ${{ runner.os }}-pnpm_store-${{ hashFiles('**/pnpm-lock.yaml') }}
          restore-keys: |
            ${{ runner.os }}-pnpm_store-

      - uses: pnpm/action-setup@v2
        with:
          version: 6.32
          run_install: true

      - run: pnpm test

      - uses: actions/upload-artifact@v3
        if: failure()
        with:
          name: failed-test-recording
          path: |
            ./cypress/videos
            ./cypress/screenshots
