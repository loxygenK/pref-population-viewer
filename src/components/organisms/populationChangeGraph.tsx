import "chart.js/auto";

import styles from "./populationChangeGraph.module.scss";
import {
  getActualMeasuredChanges,
  PopulationChange,
} from "~/domain/polulationChange";
import React from "react";
import { Prefecture } from "~/domain/prefecture";
import { useAPIProxyClient } from "~/api/proxy/hook/useAPIProxyClient";
import useSWR from "swr";
import { Graph, GraphDataSeries } from "../atom/graph";
import { AsyncDataHandler } from "../templates/asyncDataHandler";

const usePopulationChanges = (prefsToShow: Prefecture[]) => {
  const apiProxyClient = useAPIProxyClient();
  const prefIDs = prefsToShow.map((p) => p.id);

  return useSWR(["populationChange", prefIDs], () =>
    apiProxyClient.population.fetchPopulationChange(prefsToShow)
  );
};

export interface PopulationChangeGraph {
  prefsToShow: Prefecture[];
}
export const PopulationChangeGraph: React.FC<PopulationChangeGraph> = ({
  prefsToShow,
}) => {
  const { data: populationChanges, error } = usePopulationChanges(prefsToShow);

  const dataSeries = React.useMemo(() => {
    return populationChanges?.map(generateSeriesFromPopulationChanges);
  }, [populationChanges]);

  if (prefsToShow.length === 0) {
    return (
      <div className={styles.suggestion_area}>
        <span className={styles.suggestion_text}>
          都道府県リストから表示する都道府県を選択してください
        </span>
      </div>
    );
  }

  return (
    <div className={styles.graph_area}>
      <AsyncDataHandler
        data={dataSeries}
        error={error}
        skeletonProps={{
          className: styles.loading_skeleton,
        }}
      >
        {(series) => <Graph dataSeries={series} xUnit="年度" yUnit="人口数" />}
      </AsyncDataHandler>
    </div>
  );
};

const generateSeriesFromPopulationChanges = (
  populationChanges: PopulationChange
): GraphDataSeries => {
  return {
    name: populationChanges.pref.name,
    data: getActualMeasuredChanges(populationChanges).map((c) => ({
      x: c.year,
      y: c.population,
    })),
  };
};
