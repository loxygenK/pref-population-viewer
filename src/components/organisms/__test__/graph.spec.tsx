import { PopulationChangeGraph } from "../populationChangeGraph";
import { renderStatefulComponent } from "~/test/renderStatefulComponent";

import "jest-canvas-mock";
global.ResizeObserver = class MockResizeObserver {
  constructor(_callback: ResizeObserverCallback) {}

  disconnect() {}
  observe(_target: Element, _options?: ResizeObserverOptions) {}
  unobserve(_target: Element) {}
};

describe("Header component", () => {
  it("renders the suggestion correctly when no populationChange is supplied", async () => {
    const rendered = await renderStatefulComponent((root) => {
      root.render(<PopulationChangeGraph prefsToShow={[]} />);
    });

    expect(rendered).toMatchSnapshot();
  });

  it("renders the suggestion correctly when no populationChange is supplied", async () => {
    const rendered = await renderStatefulComponent((root) => {
      root.render(
        <PopulationChangeGraph
          prefsToShow={[
            { id: "0", name: "prefs-0" },
            { id: "1", name: "prefs-1" },
            { id: "2", name: "prefs-2" },
          ]}
        />
      );
    });

    expect(rendered).toMatchSnapshot();
  });
});
