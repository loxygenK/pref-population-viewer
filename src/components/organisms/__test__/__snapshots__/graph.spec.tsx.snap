// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders the suggestion correctly when no populationChange is supplied 1`] = `"<div class=\\"suggestion_area\\"><span class=\\"suggestion_text\\">都道府県リストから表示する都道府県を選択してください</span></div>"`;

exports[`Header component renders the suggestion correctly when no populationChange is supplied 2`] = `"<div class=\\"graph_area\\"><canvas role=\\"img\\" height=\\"0\\" width=\\"0\\" style=\\"display: block; box-sizing: border-box; height: 0px; width: 0px;\\"></canvas></div>"`;
